### PR TITLE
[Merged by Bors] - feat(FieldTheory/IsSepClosed): add `IsSepClosed.lift` and `IsSepClosure.equiv`

### PR DESCRIPTION
**Pull request by acmepjz**
- `IsSepClosed.lift` is a map from a separable extension `L` of `K`, into any separably closed extension `M` of `K`.

- `IsSepClosure.equiv` is a proof that any two separable closures of the
  same field are isomorphic.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by acmepjz** [comment]
Do we have maximal separable subextension work in progress?

File: Mathlib/FieldTheory/IsSepClosed.lean
Comment (jcommelin) [inline Mathlib/FieldTheory/IsSepClosed.lean:40]:
```suggestion
- In particular, a separable closure exists.
```

File: Mathlib/FieldTheory/IsSepClosed.lean
Comment (jcommelin) [inline Mathlib/FieldTheory/IsSepClosed.lean:197]: This could be a global instance, right?

Comment (acmepjz) [reply Mathlib/FieldTheory/IsSepClosed.lean:197]: Yes, what name do you suggest? `isAlgebraic_of_isSepClosure`?

Comment (acmepjz) [reply Mathlib/FieldTheory/IsSepClosed.lean:197]: But there are also a lot of instances in mathlib4 declared in their own namespaces...

Comment (riccardobrasca) [reply Mathlib/FieldTheory/IsSepClosed.lean:197]: I think Johan is proposing to make `IsSepClosure.separable` an instance, so you don't need the have line.

File: Mathlib/FieldTheory/IsAlgClosed/Basic.lean
Comment (alreadydone) [inline Mathlib/FieldTheory/IsAlgClosed/Basic.lean:313]: This is like an infinitary version of [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits) and I think `hL : Algebra.IsAlgebraic K L` can be replace by `IsIntegral K x` (or IsAlgebraic).

File: Mathlib/FieldTheory/IsSepClosed.lean
Comment (alreadydone) [inline Mathlib/FieldTheory/IsSepClosed.lean:246]: I just golfed the [splitting field version of this](https://github.com/leanprover-community/mathlib4/blob/874ff0574c19582d3d275d1c803846549cd53dce//Mathlib/FieldTheory/SplittingField/Construction.lean#L362-L365); you can probably golf this the same way, as well as the [algebraic closure version](https://github.com/leanprover-community/mathlib4/blob/874ff0574c19582d3d275d1c803846549cd53dce//Mathlib/FieldTheory/IsAlgClosed/Basic.lean#L426-L440).

File: Mathlib/FieldTheory/IsAlgClosed/Basic.lean
Comment (acmepjz) [reply Mathlib/FieldTheory/IsAlgClosed/Basic.lean:313]: > I think `hL : Algebra.IsAlgebraic K L` can be replace by `IsIntegral K x` (or IsAlgebraic).

I think so, but then we should use `IntermediateField` instead of `Subalgebra` in the structure `SubfieldWithHom`, to exclude the non-field subalgebra such as `K[X]` between `K` and `K(X)`. I think `IntermediateField` is better since we're always talking about fields here, but I'm not sure if that change will work, I'm not tried yet.

Comment (alreadydone) [reply Mathlib/FieldTheory/IsAlgClosed/Basic.lean:313]: If it's just a subalgebra, I think you can still extend the AlgHom to its field of fractions, and then you can go on to adjoin the integral/algebraic element `x`. Will indeed make the proof (which is currently super slow to work with) more complicated though; I'll look into it later.

File: Mathlib/FieldTheory/IsSepClosed.lean
Comment (acmepjz) [reply Mathlib/FieldTheory/IsSepClosed.lean:246]: Great! I golf the separable closure version. But I failed to do it for algebraic closure version because it is `L ≃ₐ[R] M` where `R` is only a commutative ring 😅 and I can't apply `algHom_bijective₂`.

File: Mathlib/FieldTheory/IsAlgClosed/Basic.lean
Comment (alreadydone) [reply Mathlib/FieldTheory/IsAlgClosed/Basic.lean:313]: Actually, we can also use [IntermediateField.algHom_mk_adjoin_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.algHom_mk_adjoin_splits), which is exactly the infinitary version of [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits) and used in the construction of [AlgHom.liftNormal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Normal.html#AlgHom.liftNormal), and has a more versatile API than `maximalSubfieldWithHom`. Maybe we should aim to remove `maximalSubfieldWithHom` in favor of the `IntermediateField.Lifts` API.

Comment (acmepjz) [reply Mathlib/FieldTheory/IsAlgClosed/Basic.lean:313]: > Actually, we can also use [IntermediateField.algHom_mk_adjoin_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Adjoin.html#IntermediateField.algHom_mk_adjoin_splits), which is exactly the infinitary version of [Polynomial.lift_of_splits](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Field.html#Polynomial.lift_of_splits), imported by this file, used in the construction of [AlgHom.liftNormal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/Normal.html#AlgHom.liftNormal), and has a more versatile API than `maximalSubfieldWithHom`. 

Yes, I think it can be used to prove `IsAlgClosed.lift` and `IsSepClosed.lift` without introducing `maximalSubfieldWithHom`.

> Maybe we should aim to remove `maximalSubfieldWithHom` in favor of the `IntermediateField.Lifts` API.
> 
> (Maybe not, it seems proofs for `maximalSubfieldWithHom` are more succinct. Maybe we should combine the best of both developments.)

Looks like that the type `IntermediateField.Lifts` is exactly the same as the type `SubfieldWithHom` but with `Subalgebra` replaced by `IntermediateField`. As I said before, I think `IntermediateField` is a better choice compared to `Subalgebra`. So I think it's fine to remove `SubfieldWithHom` completely. We can port the properties of `SubfieldWithHom` which are not mentioned for `IntermediateField.Lifts`.

As for `maximalSubfieldWithHom` and `IntermediateField.algHom_mk_adjoin_splits`, the latter merely says that there exists a map from `F(S)` to `K`, but the former says that `F(S)` is contained in `maximalSubfieldWithHom`, therefore the latter is a corollary of the former.

In fact the first line of the proof of `IntermediateField.algHom_mk_adjoin_splits`
```lean
  obtain ⟨x, hx⟩ : ∃ m : Lifts F E K, ∀ a, m ≤ a → a = m :=
    zorn_partialOrder Lifts.exists_upper_bound
```
The `x` is in fact the `maximalSubfieldWithHom`. So I think maybe we can define `maximalLift F E K` to be that element?

Comment (alreadydone) [reply Mathlib/FieldTheory/IsAlgClosed/Basic.lean:313]: Let me cc the original authors @tb65536 for the IntermediateField version [#5860](https://github.com/leanprover-community/mathlib/pull/5860), and @ChrisHughes24 / @jcommelin for the Subalgebra version [#9110](https://github.com/leanprover-community/mathlib/pull/9110). It would be possible to keep the generality of Subalgebra, and prove the maximal lift satisfies `inv_mem` to construct an IntermediateField when the source and the target are both fields, but I'm not sure how useful it would be, as both constructions currently assume the source, target and base ring are all fields. It may be straightforward to generalize to an arbitrary base ring with NoZeroSmulDivisors condition, but it seems proving [IsAlgClosed.lift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/FieldTheory/IsAlgClosed/Basic.html#IsAlgClosed.lift) (where a non-field source is allowed) still requires reducing to the field case.

**Comment by acmepjz** [comment]
I managed to modify the proofs of `IsSepClosed.lift` and `IsAlgClosed.lift`, get rid of `SubfieldWithHom`.

I think maybe the next step is port the following things for `SubfieldWithHom` to `IntermediateField.Lifts` if they are useful and not already exist for `IntermediateField.Lifts`:
- instance `LE`, `Inhabited`, `Preorder`
- definition `maximalSubfieldWithHom`
- theorem `le_def`, `compat`, `maximal_subfieldWithHom_chain_bounded`, `exists_maximal_subfieldWithHom`, `maximalSubfieldWithHom_is_maximal`
- and `mem_maximalSubfieldWithHom_of_splits`
(Note that `maximalSubfieldWithHom_eq_top` is specified for `IsAlgClosed` so it's not ported.) Then maybe we can remove `SubfieldWithHom` completely?

**Comment by alreadydone** [comment]
That's nice! I've started a new file with minimal imports, made a new definition, and adapt all lemmas about `IntermediateField.Lifts` and `SubfieldWithHom` to the new definition, with the aim to remove both eventually. The refactor shouldn't break the API so your modified proofs should work with minimal changes. Hopefully I'll submit a PR soon, but the normal closure PR will probably be delayed.

**Comment by alreadydone** [comment]
So I've experimented a bit and came to the conclusion that indeed we should stick to the IntermediateField.Lifts API and deprecate SubfieldWithHom. Some constructions for Lifts could be more elegant if more APIs are available, e.g. [Subalgebra.iSupLift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#Subalgebra.iSupLift) and [Subalgebra.coe_iSup_of_directed](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/Algebra/Subalgebra/Basic.html#Subalgebra.coe_iSup_of_directed), and defining the LE relation using `inclusion` could be better, but otherwise there isn't much to learn from the SubfieldWithHom counterpart. Since the proof of `maximalSubfieldWithHom_eq_top` is slow, it's desirable to remove it.

Among the declarations you listed:
> - instance `LE`, `Inhabited`, `Preorder`
> - definition `maximalSubfieldWithHom`
> - theorem `le_def`, `compat`, `maximal_subfieldWithHom_chain_bounded`, `exists_maximal_subfieldWithHom`, `maximalSubfieldWithHom_is_maximal`
> - and `mem_maximalSubfieldWithHom_of_splits`

+ `LE` and `Preorder` are subsumed by `PartialOrder`, and `Inhabited` subsumed by `OrderBot`.

+ `le_def` and `compat` are not important; `maximal_subfieldWithHom_chain_bounded` is basically `Lifts.exists_upper_bound` (which combines the three declarations above it)

+ definition `maximalSubfieldWithHom` also doesn't need to exist as a definition; the existence of a maximal lift is just applying Zorn's lemma to `Lifts.exists_upper_bound`, and `Lifts.exists_lift_of_splits` basically shows that every maximal lift must contain every integral element whose minimal polynomial splits in the target field and is the counterpart of `mem_maximalSubfieldWithHom_of_splits`.

Therefore, I think we could just remove `SubfieldWithHom` in another PR once this one is merged. To reduce diff, maybe you could revert the addition of `mem_maximalSubfieldWithHom_of_splits`?

**Comment by acmepjz** [comment]
> To reduce diff, maybe you could revert the addition of `mem_maximalSubfieldWithHom_of_splits`?

Sure. I've committed one.

> Since the proof of `maximalSubfieldWithHom_eq_top` is slow, it's desirable to remove it.

My attempt is to use `mem_maximalSubfieldWithHom_of_splits` to prove `maximalSubfieldWithHom_eq_top`, so that the latter is fast, but the former is slow. But since you said that:

> `Lifts.exists_lift_of_splits` basically shows that every maximal lift must contain every integral element whose minimal polynomial splits in the target field and is the counterpart of `mem_maximalSubfieldWithHom_of_splits`.

... so we can completely remove slow codes in favor of existing codes.

> * definition `maximalSubfieldWithHom` also doesn't need to exist as a definition; the existence of a maximal lift is just applying Zorn's lemma to `Lifts.exists_upper_bound`

At the expense of typing `zorn_partialOrder Lifts.exists_upper_bound`. IMHO if there is a "maximal something" in the API, it makes people easier to find them.

**Review by alreadydone — APPROVED** [review]
Thanks, LGTM!

maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
